k9mail/src/main/java/com/fsck/k9/activity/RecipientPresenter.java

```diff
@@ -0,0 +1,467 @@
+        recipientView.setRecipientExpanderVisibility(notBothAreVisible);
+    }
+
+    public void updateCryptoDisplayStatus() {
+        List<Recipient> recipients = getAllRecipients();
+
+        boolean hasNoCryptoProviderOrRecipients = cryptoProvider == null || recipients.isEmpty();
+        if (hasNoCryptoProviderOrRecipients) {
+            recipientView.hideCryptoStatus();
+            return;
+        }
+
+        if (currentCryptoMode == CryptoMode.SIGN_ONLY) {
+            recipientView.showCryptoStatus(CryptoStatusType.SIGN_ONLY);
+            return;
+        }
+        if (currentCryptoMode == CryptoMode.DISABLE) {
+            recipientView.showCryptoStatus(CryptoStatusType.DISABLED);
+            return;
+        }
+
+        boolean allKeysAvailable = true, allKeysVerified = true;
+        for (Recipient recipient : recipients) {
+            RecipientCryptoStatus cryptoStatus = recipient.getCryptoStatus();
+            if (!cryptoStatus.isAvailable()) {
+                allKeysAvailable = false;
+            } else if (cryptoStatus == RecipientCryptoStatus.AVAILABLE_UNTRUSTED) {
+                allKeysVerified = false;
+            }
+        }
+
+        if (allKeysAvailable && allKeysVerified) {
+            recipientView.showCryptoStatus(CryptoStatusType.OPPORTUNISTIC_TRUSTED);
+        } else if (allKeysAvailable) {
+            recipientView.showCryptoStatus(CryptoStatusType.OPPORTUNISTIC_UNTRUSTED);
+        } else {
+            recipientView.showCryptoStatus(CryptoStatusType.OPPORTUNISTIC_NOKEY);
+        }
+    }
+
+    @SuppressWarnings("UnusedParameters")
+    public void onToTokenAdded(Recipient recipient) {
+        updateCryptoDisplayStatus();
+    }
+    @SuppressWarnings("UnusedParameters")
+    public void onToTokenRemoved(Recipient recipient) {
+        updateCryptoDisplayStatus();
+    }
+
+    @SuppressWarnings("UnusedParameters")
+    public void onCcTokenAdded(Recipient recipient) {
+        updateCryptoDisplayStatus();
+    }
+    @SuppressWarnings("UnusedParameters")
+    public void onCcTokenRemoved(Recipient recipient) {
+        updateCryptoDisplayStatus();
+    }
+
+    @SuppressWarnings("UnusedParameters")
+    public void onBccTokenAdded(Recipient recipient) {
+        updateCryptoDisplayStatus();
+    }
+    @SuppressWarnings("UnusedParameters")
+    public void onBccTokenRemoved(Recipient recipient) {
+        updateCryptoDisplayStatus();
+    }
+
+    private void addRecipientsFromAddresses(final RecipientType recipientType, final Address... addresses) {
+        new RecipientLoader(context, cryptoProvider, addresses) {
+            @Override
+            public void deliverResult(List<Recipient> result) {
+                Recipient[] recipientArray = result.toArray(new Recipient[result.size()]);
+                recipientView.addRecipients(recipientType, recipientArray);
+                stopLoading();
+            }
+        }.startLoading();
```

Comment:
Using a loader this way feels very hacky. It's very easy to introduce a memory leak this way.